### PR TITLE
Add support for existing Cognito User Pools

### LINKED ISSUE #4207: Cannot reference existing Cognito User Pool resource in configuration
# This is a Bug Report 
(but could be a Feature Proposal if desired behaviour wasn't supposed to be supported)

## Description
I'm using serverless to manage my backend services on AWS.

I’ve divided my project into two stacks: one that defines all the lambdas, and another that defines only resources, such as DynamoDB tables, and User Pools.

I’m now stuck not knowing how to define my PreSignUp trigger in the service stack, so that it will reference a User Pool that’s already been created. I can always manually go into the User Pool console and select the correct lambda, but that just seems wrong…

I’ve tried using an ARN where pool is specified, but I get a validation error (no colons accepted for example):
```
- cognitoUserPool:
    pool: "arn:aws:cognito-idp:us-east-1:XXXX:userpool/us-east-1_XXXXX"
    trigger: PreSignUp
```

The above does not work, so I’d like to know how to do this correctly. It seems that this is designed to only work with logical names.

## Additional Data

Using Serverless Framework 1.19.0
Mac OS X 10.12.6
No stack trace
No error messages

### LINKED ISSUE #5401: The pool: <POOL_ID> property should also support Ref
_Originally posted by @HyperBrain in https://github.com/serverless/serverless/issues/3637#issuecomment-302641979_

If an existent User Pool change the logical Id to have the prefix "CognitoUserPool" a new User Pool will be created and all users will be lost.

To avoid create a new Pool Id to use the feature of Cognito Pool Triggers with Lambda Functions to keep the existent users, it's necessary have a way to reference an existent pool.

### PR DESCRIPTION
**Pull request by pmuens**
This PR adds support for existing Cognito User Pools via a custom resource.

What did you implement:

Closes #4207
Closes #5401
Refs #4241
Refs #5642
Refs #6290

/cc @exoego 

## How did you implement it:

Checks the `existing` config property of the `cognitoUserPool` event and swaps out the Cognito User Pool CloudFormation resource with a custom resource the Serverless Framework provides. The custom resource basically does raw SDK calls to update the existing Cognito User Pool. This way everything is encapsulated via CloudFormation and the stack won't get out of sync if we implement it via raw SDK calls.

## How can we verify it:

```yaml
service: test

provider:
  name: aws
  runtime: nodejs10.x

functions:
  test:
    handler: handler.hello
    events:
      - cognitoUserPool:
          pool: ExistingUserPool
          trigger: CustomMessage
          existing: true
```

## Todos:

_**Note: Run `npm run test-ci` to run all validation checks on proposed changes**_

- [x] Write tests and confirm existing functionality is not broken.  
       **Validate via `npm test`**
- [x] Write documentation
- [x] Ensure there are no lint errors.  
       **Validate via `npm run lint-updated`**  
       _Note: Some reported issues can be automatically fixed by running `npm run lint:fix`_
- [x] Ensure introduced changes match Prettier formatting.  
       **Validate via `npm run prettier-check-updated`**  
       _Note: All reported issues can be automatically fixed by running `npm run prettify-updated`_
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO

### CODE REVIEW

File: lib/plugins/aws/customResources/resources/cognitoUserPool/lib/permissions.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
+'use strict';
+
+const AWS = require('aws-sdk');
```
Comment (medikoo) [inline lib/plugins/aws/customResources/resources/cognitoUserPool/lib/permissions.js]: We may narrow require to:

```javascript
const Lambda = require('aws-sdk/clients/lambda');
```

While it won't make much difference now. It's a way to reduce size of a lambda, if packaging is setup to take only modules that are required (and I hope in near future we'll have that way integrated into core)

Comment (medikoo) [inline lib/plugins/aws/customResources/resources/cognitoUserPool/lib/permissions.js]: We may narrow require to:

```javascript
const Lambda = require('aws-sdk/clients/lambda');
```

While it won't make much difference now. It's a way to reduce size of a lambda, if packaging is setup to take only modules that are required (and I hope in near future we'll have that way integrated into core)

File: lib/plugins/aws/customResources/resources/cognitoUserPool/lib/userPool.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
+'use strict';
+
+const AWS = require('aws-sdk');
```
Comment (medikoo) [inline lib/plugins/aws/customResources/resources/cognitoUserPool/lib/userPool.js]: Same here

File: lib/plugins/aws/customResources/resources/cognitoUserPool/lib/userPool.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  const { region } = config;
+  const cognito = new AWS.CognitoIdentityServiceProvider({ region });
+
+  return findUserPoolByName(config).then(userPool =>
+    cognito
+      .describeUserPool({
+        UserPoolId: userPool.Id,
+      })
+      .promise()
+      .then(data => data)
```
Comment (medikoo) [inline lib/plugins/aws/customResources/resources/cognitoUserPool/lib/userPool.js]: This line doesn't do anything

File: lib/plugins/aws/package/compile/events/cognitoUserPool/index.js [hunk 4]
```diff
@@ -14,16 +16,172 @@ const validTriggerSources = [
...
   }
 
+  newCognitoUserPools() {
+    const { service } = this.serverless;
+    service.getAllFunctions().forEach(functionName => {
+      const functionObj = service.getFunction(functionName);
+      if (functionObj.events) {
+        functionObj.events.forEach(event => {
+          if (event.cognitoUserPool) {
+            // return immediately if it's an existing S3 event since we treat them differently
```
Comment (exoego) [inline lib/plugins/aws/package/compile/events/cognitoUserPool/index.js]: typo: S3

File: lib/plugins/aws/package/compile/events/cognitoUserPool/index.js [hunk 5]
```diff
@@ -14,16 +16,172 @@ const validTriggerSources = [
...
+      const FunctionName = functionObj.name;
+
+      if (functionObj.events) {
+        functionObj.events.forEach((event, idx) => {
+          if (
+            event.cognitoUserPool &&
+            _.isObject(event.cognitoUserPool) &&
+            event.cognitoUserPool.existing
+          ) {
+            idx = idx += 1;
```
Comment (exoego) [inline lib/plugins/aws/package/compile/events/cognitoUserPool/index.js]: Why not simply `idx++`?

File: lib/plugins/aws/package/compile/events/cognitoUserPool/index.js [hunk 6]
```diff
@@ -14,16 +16,172 @@ const validTriggerSources = [
...
+    service.getAllFunctions().forEach(functionName => {
+      let funcUsesExistingCognitoUserPool = false;
+      const functionObj = service.getFunction(functionName);
+      const FunctionName = functionObj.name;
+
+      if (functionObj.events) {
+        functionObj.events.forEach((event, idx) => {
+          if (
+            event.cognitoUserPool &&
+            _.isObject(event.cognitoUserPool) &&
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/cognitoUserPool/index.js]: This check is rudimentary

File: tests/integration-all/cognito-user-pool/service/utils.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
+'use strict';
```
Comment (medikoo) [inline tests/integration-all/cognito-user-pool/service/utils.js]: This file looks as copy of https://github.com/serverless/serverless/blob/master/tests/integration-all/s3/service/utils.js

Maybe it'll be good to find a way to reuse same file (and just ensure it in all tested services)

File: lib/plugins/aws/customResources/resources/cognitoUserPool/lib/permissions.js (hunk 1 above)
Comment (pmuens) [reply lib/plugins/aws/customResources/resources/cognitoUserPool/lib/permissions.js]: Great idea :+1:

File: lib/plugins/aws/customResources/resources/cognitoUserPool/lib/userPool.js (hunk 3 above)
Comment (pmuens) [reply lib/plugins/aws/customResources/resources/cognitoUserPool/lib/userPool.js]: Lol 🤦‍♂

File: lib/plugins/aws/package/compile/events/cognitoUserPool/index.js (hunk 6 above)
Comment (pmuens) [reply lib/plugins/aws/package/compile/events/cognitoUserPool/index.js]: 🤦‍♂

File: tests/integration-all/cognito-user-pool/service/utils.js (hunk 7 above)
Comment (pmuens) [reply tests/integration-all/cognito-user-pool/service/utils.js]: 👍 will do

**Comment by pmuens** [comment]
Thanks for the reviews @medikoo and @exoego :+1:

I just addressed your issues and updated the PR accordingly. Other than that I also added support for the usage of multiple custom resources in one service.

**Comment by pmuens** [comment]
Thanks for the review @medikoo :+1:

I agree to reduce Lodash usage and I tried to do it here. One problem I ran into was the way `Object.assign` and `_.merge` differ. `Object.assign` creates a copy whereas `_.merge` mutates the object (which is another concern). I needed a way to mutate the object so I used `_.merge` like we do in other places as well.

However I agree that we should definitely try to adhere to vanilla JS as much as possible given that we dropped support for all the old Node.js versions.

**Pull request merged**